File: kmp/composeApp/src/commonMain/kotlin/com/teya/lemonade/ColorSwatchSection.kt
package com.teya.lemonade

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp

private const val COLUMNS = 2

private val SwatchHeight: Dp = 162.dp

internal data class ColorSwatch(
    val path: String,
    val name: String,
    val fill: Color,
    val label: Color,
)

internal data class ColorSwatchGroup(
    val id: String,
    val title: String?,
    val swatches: List<ColorSwatch>,
)

@Composable
internal fun ColorSwatchSection(
    group: ColorSwatchGroup,
    outlined: Boolean = false,
) {
    Column(
        verticalArrangement = Arrangement.spacedBy(space = LemonadeTheme.spaces.spacing200),
        modifier = Modifier.padding(vertical = LemonadeTheme.spaces.spacing400),
    ) {
        if (group.title != null) {
            LemonadeUi.Text(
                text = group.title,
                textStyle = LemonadeTheme.typography.headingXXSmall,
                modifier = Modifier.padding(horizontal = LemonadeTheme.spaces.spacing100),
            )
        }
        group.swatches.chunked(size = COLUMNS).forEach { row ->
            Row(
                horizontalArrangement = Arrangement.spacedBy(space = LemonadeTheme.spaces.spacing200),
            ) {
                row.forEach { swatch ->
                    ColorSwatchBlock(
                        swatch = swatch,
                        outlined = outlined,
                        modifier = Modifier.weight(weight = 1f),
                    )
                }
                repeat(times = COLUMNS - row.size) {
                    Spacer(modifier = Modifier.weight(weight = 1f))
                }
            }
        }
    }
}

@Composable
private fun ColorSwatchBlock(
    swatch: ColorSwatch,
    outlined: Boolean,
    modifier: Modifier = Modifier,
) {
    Column(
        verticalArrangement = Arrangement.SpaceBetween,
        modifier = modifier
            .height(height = SwatchHeight)
            .background(
                color = swatch.fill,
                shape = LemonadeTheme.shapes.radius600,
            ).then(
                if (outlined) {
                    Modifier.border(
                        width = LemonadeTheme.borderWidths.base.border25,
                        color = LemonadeTheme.colors.border.borderNeutralLow,
                        shape = LemonadeTheme.shapes.radius600,
                    )
                } else {
                    Modifier
                },
            ).padding(
                horizontal = LemonadeTheme.spaces.spacing400,
                vertical = LemonadeTheme.spaces.spacing500,
            ),
    ) {
        LemonadeUi.Text(
            text = swatch.path,
            textStyle = LemonadeTheme.typography.bodyXSmallRegular,
            color = swatch.label.copy(alpha = swatch.label.alpha * LemonadeTheme.opacities.base.opacity70),
        )
        LemonadeUi.Text(
            text = swatch.name,
            textStyle = LemonadeTheme.typography.bodyXSmallMedium,
            color = swatch.label,
        )
    }
}

File: swiftui/SampleApp/ColorSwatchSection.swift
import SwiftUI
import Lemonade

struct ColorSwatch: Identifiable {
    var id: String { "\(path).\(name)" }
    let path: String
    let name: String
    let fill: Color
    let label: Color
}

/// Titles repeat across a page (every semantic group has a "Brand"), and a lazy stack
/// flattens nested `ForEach`es into one list, so the id must be unique page-wide.
struct ColorSwatchGroup: Identifiable {
    let id: String
    let title: String?
    let swatches: [ColorSwatch]
}

struct ColorSwatchSection: View {
    let group: ColorSwatchGroup
    var outlined: Bool = false

    private let columns = Array(
        repeating: GridItem(.flexible(), spacing: LemonadeTheme.spaces.spacing200),
        count: 2
    )

    var body: some View {
        VStack(alignment: .leading, spacing: LemonadeTheme.spaces.spacing200) {
            if let title = group.title {
                LemonadeUi.Text(
                    title,
                    textStyle: LemonadeTypography.shared.headingXXSmall
                )
                .padding(.horizontal, LemonadeTheme.spaces.spacing100)
            }

            LazyVGrid(columns: columns, spacing: LemonadeTheme.spaces.spacing200) {
                ForEach(group.swatches) { swatch in
                    ColorSwatchView(swatch: swatch, outlined: outlined)
                }
            }
        }
        .padding(LemonadeTheme.spaces.spacing400)
    }
}

private struct ColorSwatchView: View {
    let swatch: ColorSwatch
    let outlined: Bool

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            LemonadeUi.Text(
                swatch.path,
                textStyle: LemonadeTypography.shared.bodyXSmallRegular,
                color: swatch.label.opacity(LemonadeTheme.opacity.base.opacity70)
            )

            Spacer(minLength: 0)

            LemonadeUi.Text(
                swatch.name,
                textStyle: LemonadeTypography.shared.bodyXSmallMedium,
                color: swatch.label
            )
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .leading)
        .padding(.horizontal, LemonadeTheme.spaces.spacing400)
        .padding(.vertical, LemonadeTheme.spaces.spacing500)
        .frame(height: 162)
        .background(swatch.fill, in: LemonadeTheme.shapes.radius600)
        .overlay {
            if outlined {
                LemonadeTheme.shapes.radius600.strokeBorder(.border.borderNeutralLow, lineWidth: LemonadeTheme.borderWidth.base.border25)
            }
        }
    }
}
